lib/Transforms/IfTransform/IfTransform.cpp

```diff
@@ -0,0 +1,88 @@
+#include "lib/Transforms/IfTransform/IfTransform.h"
+
+#include "mlir/include/mlir/Dialect/SCF/IR/SCF.h"  // from @llvm-project
+#include "mlir/include/mlir/IR/Operation.h"        // from @llvm-project
+#include "mlir/include/mlir/IR/PatternMatch.h"     // from @llvm-project
+#include "mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+#include "mlir/include/mlir/Transforms/Passes.h"  // from @llvm-project
+
+using namespace mlir;
+using namespace mlir::scf;
+
+namespace mlir {
+namespace heir {
+
+#define GEN_PASS_DEF_IFTRANSFORM
+#include "lib/Transforms/IfTransform/IfTransform.h.inc"
+
+struct ConvertIfToSelect : OpRewritePattern<IfOp> {
+  using OpRewritePattern<IfOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(IfOp ifOp,
+                                PatternRewriter &rewriter) const override {
+    auto cond = ifOp.getCondition();
+    auto thenYieldArgs = ifOp.thenYield().getOperands();
+    auto elseYieldArgs = ifOp.elseYield().getOperands();
+
+    SmallVector<Value> newIfResults(ifOp->getNumResults());
+    // 1. Hoist instructions in the 'then' block
+    for (auto &operation : ifOp.getThenRegion().getOps()) {
+      auto distance = std::distance(ifOp.getThenRegion().getOps().begin(),
+                                    ifOp.getThenRegion().getOps().end());
+      if (distance == 1) {
+        break;
+      }
+      rewriter.setInsertionPointToStart(ifOp.thenBlock());
+      rewriter.moveOpBefore(&operation, ifOp);
+    }
+    // 2. Hoist instructions in the 'else' block
+    for (auto &operation : ifOp.getElseRegion().getOps()) {
+      auto distance = std::distance(ifOp.getElseRegion().getOps().begin(),
+                                    ifOp.getElseRegion().getOps().end());
+      if (distance == 1) {
+        break;
+      }
+      rewriter.setInsertionPointToStart(ifOp.thenBlock());
+      rewriter.moveOpBefore(&operation, ifOp);
+    }
```

Comment:
I think you should be able to use `llvm::make_early_inc_range` to simplify this loop. I'm actually not entirely sure the current solution is "correct" as the `moveOpBefore` might invalidate the for loop iterator (even if it works in practice), but I don't remember enough C++ stuff to know for sure ;)